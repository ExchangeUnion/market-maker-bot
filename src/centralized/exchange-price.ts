import BigNumber from 'bignumber.js';
import { Observable } from 'rxjs';
import { Config } from '../config';
import { Logger } from '../logger';

type GetPriceParams = {
  config: Config;
  logger: Logger;
};

type CentralizedExchangePriceParams = {
  config: Config;
  logger: Logger;
  getKrakenPrice$: ({
    config,
    logger,
  }: GetPriceParams) => Observable<BigNumber>;
  getBinancePrice$: ({
    config,
    logger,
  }: GetPriceParams) => Observable<BigNumber>;
};

const getCentralizedExchangePrice$ = ({
  config,
  logger,
  getKrakenPrice$,
  getBinancePrice$,
}: CentralizedExchangePriceParams): Observable<BigNumber> => {
  switch (config.CEX) {
    case 'BINANCE':
      return getBinancePrice$({ config, logger });
    case 'KRAKEN':
      return getKrakenPrice$({ config, logger });
    default:
      throw new Error(
        `Could not get price feed for unknown exchange: ${config.CEX}`
      );
  }
};

export {
  getCentralizedExchangePrice$,
  CentralizedExchangePriceParams,
  GetPriceParams,
};
