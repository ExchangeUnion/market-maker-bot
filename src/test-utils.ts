import { Config } from './config';
import { Level, Loggers } from './logger';

const testConfig = (): Config => {
  return {
    LOG_LEVEL: Level.Trace,
    CEX: 'Binance',
    CEX_API_KEY: '123',
    CEX_API_SECRET: 'abc',
    DATA_DIR: '',
    OPENDEX_CERT_PATH: `${__dirname}/../mock-data/tls.cert`,
    OPENDEX_RPC_HOST: 'localhost',
    OPENDEX_RPC_PORT: '1234',
    MARGIN: '0.06',
    BASEASSET: 'ETH',
    QUOTEASSET: 'BTC',
    CEX_BASEASSET: 'ETH',
    CEX_QUOTEASSET: 'BTC',
    TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE: '321',
    TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE: '123',
    TEST_MODE: true,
  };
};

const getLoggers = (): Loggers => {
  const mockLogger = {
    warn: () => {},
    info: () => {},
    verbose: () => {},
    debug: () => {},
    trace: () => {},
    error: () => {},
  };
  return ({
    global: mockLogger,
    centralized: mockLogger,
    opendex: mockLogger,
  } as unknown) as Loggers;
};

type TestError = {
  code: string | number;
  message: string;
};

export { getLoggers, testConfig, TestError };
