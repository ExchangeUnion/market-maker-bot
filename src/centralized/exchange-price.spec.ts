import { testConfig, getLoggers } from '../test-utils';
import { getCentralizedExchangePrice$ } from './exchange-price';

describe('getCentralizedExchangePrice$', () => {
  it('returns Binance price stream', () => {
    const config = {
      ...testConfig(),
      ...{ CEX: 'BINANCE' },
    };
    const getBinancePrice$ = jest.fn();
    const getKrakenPrice$ = jest.fn();
    const logger = getLoggers().global;
    getCentralizedExchangePrice$({
      config,
      logger,
      getBinancePrice$,
      getKrakenPrice$,
    });
    expect(getBinancePrice$).toHaveBeenCalledTimes(1);
  });

  it('returns Kraken price stream', () => {
    const config = {
      ...testConfig(),
      ...{ CEX: 'KRAKEN' },
    };
    const getBinancePrice$ = jest.fn();
    const getKrakenPrice$ = jest.fn();
    const logger = getLoggers().global;
    getCentralizedExchangePrice$({
      config,
      logger,
      getBinancePrice$,
      getKrakenPrice$,
    });
    expect(getKrakenPrice$).toHaveBeenCalledTimes(1);
  });

  it('errors for unknown exchange', () => {
    const config = {
      ...testConfig(),
      ...{ CEX: 'ASDF' },
    };
    const getBinancePrice$ = jest.fn();
    const getKrakenPrice$ = jest.fn();
    const logger = getLoggers().global;
    expect(() => {
      getCentralizedExchangePrice$({
        config,
        logger,
        getBinancePrice$,
        getKrakenPrice$,
      });
    }).toThrowErrorMatchingSnapshot();
  });
});
