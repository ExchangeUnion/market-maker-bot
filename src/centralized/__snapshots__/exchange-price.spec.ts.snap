// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCentralizedExchangePrice$ errors for unknown exchange 1`] = `"Could not get price feed for unknown exchange: ASDF"`;
